### PR TITLE
Cht tests fazzdev #1321 47

### PR DESCRIPTION
**Pull request by thefazzer**
**Cht tests fazzdev #1321 47**

- CT is at 5. Checkpoints are at 10 and 20.
- ConnectNewHeaders called with 9 new headers (from peer1).
- After that ConnectNewHeaders called with headers 5 to 15 (from peer2).
- Make sure 6 - 10 are requested for download.

**_Note: Need feedback on line @ 2320_**

Preferred: @Aprogiena @noescape00

### CODE REVIEW

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs [hunk 1]
```diff
@@ -2275,6 +2275,55 @@ public void NodeHasReorg_ReorganisedHeaders_HaveNoBlockPointers_BlockDataAvailab
...
+            ChainedHeader extendedChainTip = testContext.ExtendAChain(chainExtensionSizeOfFifteenHeaders, initialChainTip);
+            List<BlockHeader> listOfExtendedChainHeaders =
+                testContext.ChainedHeaderToList(extendedChainTip, initialChainSizeOfFiveHeaders + chainExtensionSizeOfFifteenHeaders);
+
+            // Checkpoints are at h10 and h20.
+            const int checkpoint1Height = 10;
+            const int checkpoint2Height = 20;
+
+            var checkpoint1 = new CheckpointFixture(checkpoint1Height, listOfExtendedChainHeaders[checkpoint1Height - 1]);
+            testContext.SetupCheckpoints(checkpoint1);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: Is this necessary? You call it again below

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs [hunk 2]
```diff
@@ -2275,6 +2275,55 @@ public void NodeHasReorg_ReorganisedHeaders_HaveNoBlockPointers_BlockDataAvailab
...
+            var checkpoint2 = new CheckpointFixture(checkpoint2Height, listOfExtendedChainHeaders[checkpoint2Height - 1]);
+            testContext.SetupCheckpoints(checkpoint1, checkpoint2);
+            
+            // First peer presents headers up to but excluding first checkpoint h1 -> h9.
+            List<BlockHeader> listOfBlockHeadersOneToNine = listOfExtendedChainHeaders.Take(9).ToList();
+            chainedHeaderTree.ConnectNewHeaders(peerOneId, listOfBlockHeadersOneToNine);
+
+            // Second peer presents headers including checkpoint1, excluding checkpoint2: h5 -> 15.
+            List<BlockHeader> listOfBlockHeadersFiveToFifteen = listOfExtendedChainHeaders.GetRange(initialChainSizeOfFiveHeaders, 10);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: The comment says 5 to 15, but it is goes only to 14

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs [hunk 3]
```diff
@@ -2275,6 +2275,55 @@ public void NodeHasReorg_ReorganisedHeaders_HaveNoBlockPointers_BlockDataAvailab
...
+            ConnectNewHeadersResult connectNewHeadersResult = chainedHeaderTree.ConnectNewHeaders(peerTwoId, listOfBlockHeadersFiveToFifteen);
+
+            // Headers h6 -> h10 should be marked for download.
+            connectNewHeadersResult.DownloadFrom.HashBlock.Should().Be(extendedChainTip.GetAncestor(initialChainSizeOfFiveHeaders + 1).HashBlock); // h6
+            connectNewHeadersResult.DownloadTo.HashBlock.Should().Be(extendedChainTip.GetAncestor(checkpoint1Height).HashBlock); // h10
+            connectNewHeadersResult.HaveBlockDataAvailabilityStateOf(BlockDataAvailabilityState.BlockRequired).Should().BeTrue(); 
+        }
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: Can we also check the rest of the chain from 11 to 20 that its availability state is header only and not requested?

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs (hunk 1 above)
Comment (thefazzer) [reply src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: Removed.

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs (hunk 2 above)
Comment (thefazzer) [reply src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: Fixed.

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs (hunk 3 above)
Comment (thefazzer) [reply src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: Added this check.

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs [hunk 4]
```diff
@@ -2275,6 +2275,62 @@ public void NodeHasReorg_ReorganisedHeaders_HaveNoBlockPointers_BlockDataAvailab
...
+        /// <summary>
+        /// Issue 47 @ CT is at 5. Checkpoints are at 10 and 20.
+        /// ConnectNewHeaders called with 9 new headers (from peer1).
+        /// After that ConnectNewHeaders called with headers 5 to 15 (from peer2).
+        /// Make sure 6 - 10 are requested for download.
+        /// </summary>
+        [Fact]
+        public void PresentHeaders_CheckpointsEnabledAndSet_PresentHeadersFromAlternatePeer_MarkedForDownload()
+        {
+            const int initialChainSizeOfFiveHeaders= 5;
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: Space before = is missing

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs [hunk 5]
```diff
@@ -2275,6 +2275,62 @@ public void NodeHasReorg_ReorganisedHeaders_HaveNoBlockPointers_BlockDataAvailab
...
+            var checkpoint1 = new CheckpointFixture(checkpoint1Height, listOfExtendedChainHeaders[checkpoint1Height - 1]);
+            var checkpoint2 = new CheckpointFixture(checkpoint2Height, listOfExtendedChainHeaders[checkpoint2Height - 1]);
+            testContext.SetupCheckpoints(checkpoint1, checkpoint2);
+            
+            // First peer presents headers up to but excluding first checkpoint h1 -> h9.
+            List<BlockHeader> listOfBlockHeadersOneToNine = listOfExtendedChainHeaders.Take(9).ToList();
+            chainedHeaderTree.ConnectNewHeaders(peerOneId, listOfBlockHeadersOneToNine);
+
+            // Second peer presents headers including checkpoint1, excluding checkpoint2: h5 -> h15.
+            List<BlockHeader> listOfBlockHeadersFiveToFifteen = listOfExtendedChainHeaders.GetRange(initialChainSizeOfFiveHeaders, 11);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: Shouldn't the first argument of GetRange have -1?

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs (hunk 4 above)
Comment (thefazzer) [reply src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: Fixed.

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs (hunk 5 above)
Comment (thefazzer) [reply src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: Correct - it was actually presenting h6 -> h16.
Fixed.

**Comment by Aprogiena** [comment]
Looks good

**Pull request merged**